edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Diagnostics/util/PrometheusMetricsParser.cs

```diff
@@ -13,51 +13,135 @@ namespace Microsoft.Azure.Devices.Edge.Agent.Diagnostics.Util
+                            $"{Whitespace}*" +
+                            Label +
+                        @")*" +
+                        @"(?:" +
+                            @"," +
+                            $"{Whitespace}*" +
+                        @")?" +
+                    @")?" +
+                    @"\}" +
+                    $"{Whitespace}*" +
+                @")?" +
+                MetricValue +
+                @"(?:" +
+                    $"{Whitespace}+" +
+                    MetricTimestamp +
+                @")?" +
+            @"$",
             RegexOptions.Compiled);
 
+        private static readonly DateTime MetricTimestampEpoch = DateTime.UnixEpoch;
+
         public static IEnumerable<Metric> ParseMetrics(DateTime timeGeneratedUtc, string prometheusMessage)
         {
             Preconditions.CheckArgument(timeGeneratedUtc.Kind == DateTimeKind.Utc, $"Metric {nameof(timeGeneratedUtc)} parameter only supports in UTC.");
             using (StringReader sr = new StringReader(prometheusMessage))
             {
                 string line;
-                while ((line = sr.ReadLine()?.Trim()) != null)
+                while ((line = sr.ReadLine()) != null)
                 {
                     if (line.Length == 0 || line[0] == '#')
                     {
                         continue;
                     }
 
                     Match match = PrometheusSchemaRegex.Match(line);
-                    if (match.Success)
+                    if (!match.Success)
+                    {
+                        continue;
+                    }
+
+                    double metricValue;
+                    if (!double.TryParse(match.Groups["metricvalue"]?.Value, out metricValue))
                     {
-                        double metricValue;
-                        if (!double.TryParse(match.Groups["metricvalue"]?.Value, out metricValue))
+                        continue;
+                    }
+
+                    string metricName = match.Groups["metricname"]?.Value ?? string.Empty;
+                    var tagNames = (match.Groups["labelname"].Captures as IEnumerable<Capture>).Select(c => c.Value);
+                    var tagValues =
+                        (match.Groups["labelvalue"].Captures as IEnumerable<Capture>)
+                        .Select(value =>
                         {
-                            continue;
-                        }
+                            // Evaluate escaped characters
+                            string result = string.Empty;
+                            bool escaped = false;
+                            foreach (char c in value.Value)
+                            {
+                                if (escaped)
+                                {
+                                    if (c == 'n')
+                                    {
+                                        result += "\n";
+                                    }
```

Comment:
Do we have to handle other special characters according to prometheus spec? Seems like this approach could be error prone.

Reply:
The spec (and golang parser) only all `\\`, `\"` and `\n` escapes. With the exception of `\n`, the rest are just `result += c;`, so I didn't write explicit code for them. The regex already enforces that only those three escapes will be accepted anyway.